libraries/Microsoft.Bot.Builder.Dialogs/ComponentDialog.cs

```diff
@@ -120,17 +108,48 @@ protected virtual async Task<DialogTurnResult> OnDialogBeginAsync(DialogContext
             return await dc.BeginAsync(InitialDialogId, options, cancellationToken).ConfigureAwait(false);
         }
 
-        protected virtual async Task OnDialogEndAsync(DialogContext dc, DialogReason reason, CancellationToken cancellationToken = default(CancellationToken))
+        protected virtual async Task<DialogTurnResult> OnDialogEndAsync(DialogContext dc, DialogReason reason, CancellationToken cancellationToken = default(CancellationToken))
         {
             if (reason == DialogReason.CancelCalled)
             {
                 await dc.CancelAllAsync(cancellationToken).ConfigureAwait(false);
+                return new DialogTurnResult(DialogStatus.Cancelled);
+            }
+
+            if (dc.ActiveDialog != null)
+            {
+                await this.OnDialogRepromptAsync(dc).ConfigureAwait(false);
+                return Dialog.EndOfTurn;
+            }
+            else
+            {
+                // LAMIL: Why doesn't end follow the same pattern as all the other methods?
+                return new DialogTurnResult(DialogStatus.Complete);
             }
         }
 
         protected virtual async Task<DialogTurnResult> OnDialogContinueAsync(DialogContext dc, CancellationToken cancellationToken = default(CancellationToken))
         {
-            return await dc.ContinueAsync(cancellationToken).ConfigureAwait(false);
+            var result = await dc.ContinueAsync(cancellationToken).ConfigureAwait(false);
+
+            switch (result.Status)
+            {
+                case DialogStatus.Complete:
+                    {
+                        return await dc.EndAsync(result.Result).ConfigureAwait(false);
+                    }
+
+                case DialogStatus.Cancelled:
```

Comment:
>case DialogStatus.Cancelled [](start = 16, length = 27)

I need to better understand what you're doing here. This doesn't seem right to me.